pbft/pbft_checkpoint_manager.cpp

```diff
@@ -0,0 +1,323 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+#include <pbft/pbft_checkpoint_manager.hpp>
+#include <proto/pbft.pb.h>
+#include <utils/bytes_to_debug_string.hpp>
+#include <utils/make_endpoint.hpp>
+#include <algorithm>
+#include <pbft/pbft.hpp>
+#include <iterator>
+
+namespace
+{
+    const std::chrono::seconds CHECKPOINT_CATCHUP_GRACE_PERIOD{std::chrono::seconds(30)};
+}
+
+using namespace bzn;
+
+pbft_checkpoint_manager::pbft_checkpoint_manager(
+        std::shared_ptr<asio::io_context_base> io_context,
+        std::shared_ptr<storage_base> storage,
+        std::shared_ptr<pbft_config_store> config_store,
+        std::shared_ptr<node_base> node
+)
+        : io_context(std::move(io_context))
+        , storage(std::move(storage))
+        , config_store(std::move(config_store))
+        , node(std::move(node))
+{
+    this->init_persists();
+}
+
+void
+pbft_checkpoint_manager::start()
+{
+    this->node->register_for_message(bzn_envelope::kCheckpointMsg,
+            std::bind(&pbft_checkpoint_manager::handle_checkpoint_message, shared_from_this(), std::placeholders::_1, std::placeholders::_2));
+}
+
+checkpoint_t
+pbft_checkpoint_manager::get_latest_stable_checkpoint() const
+{
+    std::lock_guard<std::mutex> lock(this->lock);
+    return this->latest_stable_checkpoint.value();
+}
+
+checkpoint_t
+pbft_checkpoint_manager::get_latest_local_checkpoint() const
+{
+    std::lock_guard<std::mutex> lock(this->lock);
+    return this->latest_local_checkpoint.value();
+}
+
+void
+pbft_checkpoint_manager::local_checkpoint_reached(const bzn::checkpoint_t& cp)
+{
+    std::lock_guard<std::mutex> lock(this->lock);
+
+    LOG(info) << "reached local checkpoint at " << cp.first << " (" << bytes_to_debug_string(cp.second) << ")";
+    if (cp.first <= this->latest_local_checkpoint.value().first)
+    {
+        throw std::runtime_error("reached a checkpoint at " + std::to_string(cp.first)
+                                 + " which is not newer than the previous checkpoint at "
+                                 + std::to_string(this->latest_local_checkpoint.value().first));
+    }
+
+    this->latest_local_checkpoint = cp;
+
+    // Compose and send our own cp message
+    checkpoint_msg cp_msg;
+    cp_msg.set_sequence(cp.first);
+    cp_msg.set_state_hash(cp.second);
+
+    bzn_envelope msg;
+    msg.set_checkpoint_msg(cp_msg.SerializeAsString());
+
+    auto msg_ptr = std::make_shared<bzn_envelope>(msg);
+    for (const auto& peer : *(this->config_store->current()->get_peers()))
+    {
+        this->node->send_signed_message(make_endpoint(peer), msg_ptr);
+    }
+
+    if (cp.first == this->latest_local_checkpoint.value().first && cp.second != this->latest_local_checkpoint.value().second)
+    {
```

Comment:
KEP-1321 cool, verifying that the local checkpoint matches the swarm, sending a state request if it does not...